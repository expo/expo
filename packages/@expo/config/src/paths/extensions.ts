import assert from 'assert';

export type LanguageOptions = {
  isTS: boolean;
  isModern: boolean;
  isReact: boolean;
};

export function getExtensions(
  platforms: string[],
  extensions: string[],
  workflows: string[]
): string[] {
  // In the past we used spread operators to collect the values so now we enforce type safety on them.
  assert(Array.isArray(platforms), 'Expected: `platforms: string[]`');
  assert(Array.isArray(extensions), 'Expected: `extensions: string[]`');
  assert(Array.isArray(workflows), 'Expected: `workflows: string[]`');

  const fileExtensions = [];
  // support .expo files
  for (const workflow of [...workflows, '']) {
    // Ensure order is correct: [platformA.js, platformB.js, js]
    for (const platform of [...platforms, '']) {
      // Support both TypeScript and JavaScript
      for (const extension of extensions) {
        fileExtensions.push([platform, workflow, extension].filter(Boolean).join('.'));
      }
    }
  }
  return fileExtensions;
}

export function getLanguageExtensionsInOrder({
  isTS,
  isModern,
  isReact,
}: LanguageOptions): string[] {
  // @ts-ignore: filter removes false type
  const addLanguage = (lang: string): string[] => [lang, isReact && `${lang}x`].filter(Boolean);

  // Support JavaScript
  let extensions = addLanguage('js');

  if (isModern) {
    extensions.unshift('mjs');
  }
  if (isTS) {
    extensions = [...addLanguage('ts'), ...extensions];
  }

  return extensions;
}

export function getBareExtensions(
  platforms: string[],
  languageOptions: LanguageOptions = { isTS: true, isModern: true, isReact: true }
): string[] {
  const fileExtensions = getExtensions(
    platforms,
    getLanguageExtensionsInOrder(languageOptions),
    []
  );
  // Always add these last
  _addMiscellaneousExtensions(platforms, fileExtensions);
  return fileExtensions;
}

function _addMiscellaneousExtensions(platforms: string[], fileExtensions: string[]): string[] {
  // Always add these with no platform extension
  // In the future we may want to add platform and workspace extensions to json.
  fileExtensions.push('json');
  // Native doesn't currently support web assembly.
  if (platforms.includes('web')) {
    fileExtensions.push('wasm');
  }
  return fileExtensions;
}
