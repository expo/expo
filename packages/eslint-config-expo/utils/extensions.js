const jsExtensions = ['.js', '.jsx'];
const tsExtensions = ['.ts', '.tsx', '.d.ts'];

const platformSubextensions = ['.android', '.ios', '.web', '.native'];

function computeExpoExtensions(baseExtensions, platformSubextensions) {
  const expoExtensions = [];
  for (const platform of [...platformSubextensions, '']) {
    for (const base of baseExtensions) {
      expoExtensions.push(`${platform}${base}`);
    }
  }
  return expoExtensions;
}

module.exports = {
  jsExtensions,
  tsExtensions,
  platformSubextensions,
  computeExpoExtensions,
};
