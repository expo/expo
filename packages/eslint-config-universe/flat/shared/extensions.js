const jsExtensions = ['.js', '.jsx'];
const tsExtensions = ['.ts', '.tsx', '.d.ts'];

const platformSubextensions = ['.android', '.ios', '.web', '.native'];

function computeExpoExtensions(baseExtensions, platformSubextensions) {
  const expoExtensions = [];
  for (const expo of ['.expo', '']) {
    for (const platform of [...platformSubextensions, '']) {
      for (const base of baseExtensions) {
        expoExtensions.push(`${expo}${platform}${base}`);
      }
    }
  }
  return expoExtensions;
}

module.exports = {
  jsExtensions,
  tsExtensions,
  platformSubextensions,
  computeExpoExtensions,
};
