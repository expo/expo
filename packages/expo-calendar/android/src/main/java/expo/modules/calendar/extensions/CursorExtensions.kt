package expo.modules.calendar.extensions

import android.database.Cursor
import android.util.Log
import expo.modules.calendar.CalendarModule
import expo.modules.calendar.exceptions.ColumnMissingException
import kotlin.jvm.Throws

fun Cursor.getOptionalString(columnName: String): String? {
  return getColumnIndex(columnName)
    .takeIf { it >= 0 }
    ?.let { getString(it) }
}

@Throws(ColumnMissingException::class)
fun Cursor.getRequiredString(columnName: String): String {
  return getOptionalString(columnName)
    ?: throw ColumnMissingException("String value not found for column `$columnName`")
}

fun Cursor.getIntOrDefault(columnName: String, default: Int = 0): Int {
  return getColumnIndex(columnName)
    .takeIf { it >= 0 }
    ?.let { getInt(it) }
    ?: default
}

fun <T> Cursor.getValueList(columnName: String, mapper: (v: Int) -> T): List<T> =
  getRequiredString(columnName)
    .split(",")
    .mapNotNull { v ->
      try {
        mapper(v.toInt())
      } catch (e: NumberFormatException) {
        Log.e(CalendarModule.TAG, "Couldn't convert value constant into an int.", e)
        null
      }
    }
    .toList()
