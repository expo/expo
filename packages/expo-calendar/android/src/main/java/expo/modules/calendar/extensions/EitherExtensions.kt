package expo.modules.calendar.extensions

import expo.modules.calendar.CalendarUtils
import expo.modules.calendar.exceptions.DateParseException
import expo.modules.kotlin.apifeatures.EitherType
import expo.modules.kotlin.types.Either
import java.text.ParseException
import java.util.Calendar

// Represents date/time input from JS. Can be either string or number.
@OptIn(EitherType::class)
typealias DateTimeInput = Either<String, Long>

// Returns Unix timestamp in milliseconds.
@OptIn(EitherType::class)
fun DateTimeInput.getTimeInMillis(): Long? {
  if (this.`is`(Long::class)) {
    return this.second()
  }

  try {
    val parsedDate = CalendarUtils.sdf.parse(this.first())
    if (parsedDate == null) {
      throw DateParseException("Parsed date for string '${this.first()}' is null")
    }
    val cal = Calendar.getInstance().apply { time = parsedDate }
    return cal.timeInMillis
  } catch (e: ParseException) {
    throw DateParseException("Date '${this.first()}' could not be parsed", e)
  }
}
