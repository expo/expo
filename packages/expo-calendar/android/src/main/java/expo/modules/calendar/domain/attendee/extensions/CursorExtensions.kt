package expo.modules.calendar.domain.attendee.extensions

import android.database.Cursor
import android.provider.CalendarContract
import expo.modules.calendar.domain.attendee.enums.AttendeeRole
import expo.modules.calendar.domain.attendee.enums.AttendeeStatus
import expo.modules.calendar.domain.attendee.enums.AttendeeType
import expo.modules.calendar.domain.attendee.records.Attendee
import expo.modules.calendar.extensions.getIntOrDefault
import expo.modules.calendar.extensions.getOptionalString
import expo.modules.calendar.extensions.getRequiredString

fun Cursor.extractAttendee(): Attendee {
  val roleValue = getIntOrDefault(CalendarContract.Attendees.ATTENDEE_RELATIONSHIP)
  val typeValue = getIntOrDefault(CalendarContract.Attendees.ATTENDEE_TYPE)
  val statusValue = getIntOrDefault(CalendarContract.Attendees.ATTENDEE_STATUS)
  return Attendee(
    id = getOptionalString(CalendarContract.Attendees._ID),
    name = getRequiredString(CalendarContract.Attendees.ATTENDEE_NAME),
    email = getOptionalString(CalendarContract.Attendees.ATTENDEE_EMAIL),
    role = AttendeeRole.fromContentProviderValue(roleValue),
    type = AttendeeType.fromContentProviderValue(typeValue),
    status = AttendeeStatus.fromContentProviderValue(statusValue)
  )
}
