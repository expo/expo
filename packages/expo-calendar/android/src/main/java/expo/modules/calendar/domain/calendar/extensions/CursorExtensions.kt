package expo.modules.calendar.domain.calendar.extensions

import android.database.Cursor
import android.provider.CalendarContract
import expo.modules.calendar.domain.attendee.enums.AttendeeType
import expo.modules.calendar.domain.calendar.enums.CalendarAccessLevel
import expo.modules.calendar.domain.calendar.records.CalendarEntity
import expo.modules.calendar.domain.calendar.records.CalendarSource
import expo.modules.calendar.domain.event.enums.AlarmMethod
import expo.modules.calendar.domain.event.enums.Availability
import expo.modules.calendar.extensions.getIntOrDefault
import expo.modules.calendar.extensions.getOptionalString
import expo.modules.calendar.extensions.getValueList

fun Cursor.extractCalendar(): CalendarEntity {
  val accessLevel = getIntOrDefault(CalendarContract.Calendars.CALENDAR_ACCESS_LEVEL)
  val color = String.format("#%06X", 0xFFFFFF and getIntOrDefault(CalendarContract.Calendars.CALENDAR_COLOR))

  val allowedAvailabilities = getValueList(CalendarContract.Calendars.ALLOWED_AVAILABILITY) { Availability.fromContentProviderValue(it) }
  val allowedReminders = getValueList(CalendarContract.Calendars.ALLOWED_REMINDERS) { AlarmMethod.fromContentProviderValue(it) }
  val allowedAttendeeTypes = getValueList(CalendarContract.Calendars.ALLOWED_ATTENDEE_TYPES) { AttendeeType.fromContentProviderValue(it) }

  return CalendarEntity(
    id = getOptionalString(CalendarContract.Calendars._ID),
    name = getOptionalString(CalendarContract.Calendars.NAME),
    title = getOptionalString(CalendarContract.Calendars.CALENDAR_DISPLAY_NAME),
    isPrimary = getIntOrDefault(CalendarContract.Calendars.IS_PRIMARY) == 1,
    color = color,
    ownerAccount = getOptionalString(CalendarContract.Calendars.OWNER_ACCOUNT),
    timeZone = getOptionalString(CalendarContract.Calendars.CALENDAR_TIME_ZONE),
    allowedAvailabilities = allowedAvailabilities,
    allowedReminders = allowedReminders,
    allowedAttendeeTypes = allowedAttendeeTypes,
    isVisible = getIntOrDefault(CalendarContract.Calendars.VISIBLE) != 0,
    isSynced = getIntOrDefault(CalendarContract.Calendars.SYNC_EVENTS) != 0,
    accessLevel = CalendarAccessLevel.fromContentProviderValue(accessLevel),
    allowsModifications = accessLevel == CalendarContract.Calendars.CAL_ACCESS_ROOT ||
      accessLevel == CalendarContract.Calendars.CAL_ACCESS_OWNER ||
      accessLevel == CalendarContract.Calendars.CAL_ACCESS_EDITOR ||
      accessLevel == CalendarContract.Calendars.CAL_ACCESS_CONTRIBUTOR,
    source = CalendarSource(
      name = getOptionalString(CalendarContract.Calendars.ACCOUNT_NAME),
      type = getOptionalString(CalendarContract.Calendars.ACCOUNT_TYPE)
    )
  )
}
