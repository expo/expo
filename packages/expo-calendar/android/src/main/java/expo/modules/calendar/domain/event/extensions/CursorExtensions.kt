package expo.modules.calendar.domain.event.extensions

import android.content.ContentResolver
import android.database.Cursor
import android.provider.CalendarContract
import expo.modules.calendar.CalendarUtils
import expo.modules.calendar.domain.event.enums.AlarmMethod
import expo.modules.calendar.domain.event.enums.Availability
import expo.modules.calendar.domain.event.enums.EventAccessLevel
import expo.modules.calendar.domain.event.records.Alarm
import expo.modules.calendar.domain.event.records.EventEntity
import expo.modules.calendar.domain.event.records.RecurrenceRuleEntity
import expo.modules.calendar.extensions.getIntOrDefault
import expo.modules.calendar.extensions.getOptionalString
import java.util.Calendar

fun Cursor.extractAlarm(): Alarm {
  val minutesIndex = getColumnIndexOrThrow(CalendarContract.Reminders.MINUTES)
  val methodIndex = getColumnIndexOrThrow(CalendarContract.Reminders.METHOD)

  return Alarm(
    relativeOffset = -getInt(minutesIndex),
    method = AlarmMethod.fromContentProviderValue(getInt(methodIndex))
  )
}

fun Cursor.extractEvent(contentResolver: ContentResolver): EventEntity {
  // may be CalendarContract.Instances.BEGIN or CalendarContract.Events.DTSTART (which have different string values)
  val startDate = getString(3)?.let {
    val foundStartDate = Calendar.getInstance().apply { timeInMillis = it.toLong() }
    CalendarUtils.sdf.format(foundStartDate.time)
  } ?: ""

  // may be CalendarContract.Instances.END or CalendarContract.Events.DTEND (which have different string values)
  val endDate = getString(4)?.let {
    val foundEndDate = Calendar.getInstance().apply { timeInMillis = it.toLong() }
    CalendarUtils.sdf.format(foundEndDate.time)
  } ?: ""

  val availabilityValue = getIntOrDefault(CalendarContract.Events.AVAILABILITY)
  val accessLeveLValue = getIntOrDefault(CalendarContract.Events.ACCESS_LEVEL)
  val instanceId = takeIf { it.columnCount > 18 }?.getString(18)

  val recurrenceRule = getOptionalString(CalendarContract.Events.RRULE)?.let {
    RecurrenceRuleEntity.fromRuleString(it)
  }

  val eventID = getString(0)
  return EventEntity(
    id = eventID,
    calendarId = getOptionalString(CalendarContract.Events.CALENDAR_ID),
    title = getOptionalString(CalendarContract.Events.TITLE),
    notes = getOptionalString(CalendarContract.Events.DESCRIPTION),
    startDate = startDate,
    endDate = endDate,
    alarms = getAlarmsForEvent(contentResolver, eventID.toLong()),
    availability = Availability.fromContentProviderValue(availabilityValue),
    allDay = getIntOrDefault(CalendarContract.Events.ALL_DAY) != 0,
    location = getOptionalString(CalendarContract.Events.EVENT_LOCATION),
    organizerEmail = getOptionalString(CalendarContract.Events.ORGANIZER),
    guestsCanModify = getIntOrDefault(CalendarContract.Events.GUESTS_CAN_MODIFY) != 0,
    guestsCanInviteOthers = getIntOrDefault(CalendarContract.Events.GUESTS_CAN_INVITE_OTHERS) != 0,
    guestsCanSeeGuests = getIntOrDefault(CalendarContract.Events.GUESTS_CAN_SEE_GUESTS) != 0,
    timeZone = getOptionalString(CalendarContract.Events.EVENT_TIMEZONE),
    endTimeZone = getOptionalString(CalendarContract.Events.EVENT_END_TIMEZONE),
    accessLevel = EventAccessLevel.fromContentProviderValue(accessLeveLValue),
    recurrenceRule = recurrenceRule,
    originalId = getOptionalString(CalendarContract.Events.ORIGINAL_ID),
    instanceId = instanceId
  )
}

private fun getAlarmsForEvent(contentResolver: ContentResolver, eventID: Long): List<Alarm> {
  val alarms = mutableListOf<Alarm>()
  val cursor = CalendarContract.Reminders.query(
    contentResolver,
    eventID,
    arrayOf(
      CalendarContract.Reminders.MINUTES,
      CalendarContract.Reminders.METHOD
    )
  )
  while (cursor.moveToNext()) {
    alarms.add(cursor.extractAlarm())
  }
  return alarms
}
