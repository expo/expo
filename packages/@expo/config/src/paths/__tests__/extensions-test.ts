import { getExtensions, getBareExtensions, getLanguageExtensionsInOrder } from '../extensions';

describe(getExtensions, () => {
  it(`enforces \`string[]\``, async () => {
    // @ts-expect-error: test invalid input
    expect(() => getExtensions('web')).toThrow(`string[]`);
    // @ts-expect-error: test invalid input
    expect(() => getExtensions([], 'js')).toThrow(`string[]`);
    // @ts-expect-error: test invalid input
    expect(() => getExtensions([], [], 'expo')).toThrow(`string[]`);
    // @ts-expect-error: test invalid input
    expect(() => getExtensions([], null, [])).toThrow(`string[]`);
  });
});

describe(getBareExtensions, () => {
  it(`creates extensions for web`, async () => {
    expect(getBareExtensions(['web'])).toMatchSnapshot();
  });

  it(`creates extensions for iOS`, async () => {
    expect(getBareExtensions(['ios', 'native'])).toMatchSnapshot();
  });
});

// Enforce that all extensions are returned in the correct order, this is very important!
describe(getLanguageExtensionsInOrder, () => {
  // Return only support for the bare minimum
  it(`JS only`, async () => {
    expect(
      getLanguageExtensionsInOrder({ isTS: false, isModern: false, isReact: false })
    ).toStrictEqual(['js']);
  });

  // Modern should come before JS in a basic MJS project
  it(`Modern JS`, async () => {
    expect(
      getLanguageExtensionsInOrder({ isTS: false, isModern: true, isReact: false })
    ).toStrictEqual(['mjs', 'js']);
  });

  // TypeScript should precede JavaScript
  it(`TypeScript`, async () => {
    expect(
      getLanguageExtensionsInOrder({ isTS: true, isModern: false, isReact: false })
    ).toStrictEqual(['ts', 'js']);
  });

  // JS should precede JSX files
  it(`React`, async () => {
    expect(
      getLanguageExtensionsInOrder({ isTS: false, isModern: false, isReact: true })
    ).toStrictEqual(['js', 'jsx']);
  });

  // Modern should come before JS in a basic MJS project
  it(`TypeScript React`, async () => {
    expect(
      getLanguageExtensionsInOrder({ isTS: true, isModern: false, isReact: true })
    ).toStrictEqual(['ts', 'tsx', 'js', 'jsx']);
  });

  // Should include all extensions in the correct order
  it(`Modern TypeScript React`, async () => {
    expect(
      getLanguageExtensionsInOrder({ isTS: true, isModern: true, isReact: true })
    ).toStrictEqual(['ts', 'tsx', 'mjs', 'js', 'jsx']);
  });
});
