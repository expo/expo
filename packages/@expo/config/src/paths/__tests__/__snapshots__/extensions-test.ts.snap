// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getBareExtensions creates extensions for iOS 1`] = `
[
  "ios.ts",
  "ios.tsx",
  "ios.mjs",
  "ios.js",
  "ios.jsx",
  "native.ts",
  "native.tsx",
  "native.mjs",
  "native.js",
  "native.jsx",
  "ts",
  "tsx",
  "mjs",
  "js",
  "jsx",
  "json",
]
`;

exports[`getBareExtensions creates extensions for web 1`] = `
[
  "web.ts",
  "web.tsx",
  "web.mjs",
  "web.js",
  "web.jsx",
  "ts",
  "tsx",
  "mjs",
  "js",
  "jsx",
  "json",
  "wasm",
]
`;
